File: src/com/craftinginterpreters/tool/GenerateAst.java
package com.craftinginterpreters.tool;

import java.awt.*;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;

public class GenerateAst {
    public static void main(String[] args) throws IOException {
//        if (args.length != 1) {
//            System.err.println("Usage: generate_ast <output directory>");
//            System.exit(64);
//        }
        String outputdir = "C:\\dev\\javaproj\\jlox\\src\\com\\craftinginterpreters\\lox";
        defineAst(outputdir, "Expr", Arrays.asList(
                "Binary   : Expr left, Token operator, Expr right",
                "Grouping : Expr expression",
                "Literal  : Object value",
                "Unary    : Token operator, Expr right"
        ));

    }

    private static void defineAst(
            String outputDir, String baseName, List<String> types)
            throws IOException {
        String path = outputDir + '\\' + baseName + ".java";
        PrintWriter writer = new PrintWriter(path, StandardCharsets.UTF_8);

        writer.println("package com.craftinginterpreters.lox;");
        writer.println();
        writer.println("import java.util.List;");
        writer.println();
        writer.println("abstract class " + baseName + " {");

        defineVisitor(writer, baseName, types);
        // The AST classes
        for (String type: types) {
            String ClassName = type.split(":")[0].trim();
            String fields = type.split(":")[1].trim();
            defineType(writer, baseName, ClassName, fields);
        }
        // The base accept method.
        writer.println();
        writer.println("        abstract <R> R accept(Visitor<R> visitor);");
        writer.println("}");
        writer.close();
    }

    private static void defineType(
            PrintWriter writer, String baseName,
            String className, String fieldList) {
        writer.println("    static class " + className + " extends " + baseName + " {");
        //Constructor
        writer.println("        " + className + "(" + fieldList + ") {");

        //Store parameters in the field
        String [] fields = fieldList.split(", ");
        for (String field: fields) {
            String name = field.split(" ")[1];
            writer.println("            this." + name + " = " + name + ";");
        }

        writer.println("        }");

        // Visitor Pattern
        writer.println();
        writer.println("        @Override");
        writer.println("        <R> R accept(Visitor<R> visitor) {");
        writer.println("            return visitor.visit" + className +
                baseName + "(this);");
        writer.println("        }");

        //Fields.
        writer.println();
        for (String field: fields) {
            writer.println("        final " + field + ";");
        }

        writer.println("    }");
    }

    private static void defineVisitor(
        PrintWriter writer, String baseName, List<String> types) {
        writer.println("    interface Visitor<R> {");

        for (String type: types) {
            String typeName = type.split(":")[0].trim();
            writer.println("        R visit" + typeName + baseName + "(" +
                    typeName + " " + baseName.toLowerCase() + ");");
        }

        writer.println("    }");
    }



}

File: src/com/craftinginterpreters/tool/GenerateAst.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <filesystem>
#include <sstream>

using namespace std;
namespace fs = std::filesystem;

void defineAst(const string& outputDir, const string& baseName, const vector<string>& types);
void defineType(ofstream& writer, const string& baseName, const string& className, const string& fieldList);
void defineVisitor(ofstream& writer, const string& baseName, const vector<string>& types);

int main() {
    string outputDir = "C:\\dev\\javaproj\\jlox\\src\\com\\craftinginterpreters\\lox";

    defineAst(outputDir, "Expr", {
        "Assign   : Token name, Expr value",
        "Binary   : Expr left, Token operator, Expr right",
        "Grouping : Expr expression",
        "Literal  : Object value",
        "Unary    : Token operator, Expr right",
        "Variable : Token name"
    });

    defineAst(outputDir, "Stmt", {
        "Block      : List<Stmt> statements",
        "Expression : Expr expression",
        "If         : Expr condition, Stmt thenBranch, Stmt else_Branch",
        "Print      : Expr expression",
        "Var        : Token name, Expr initializer"
    });

    return 0;
}

void defineAst(const string& outputDir, const string& baseName, const vector<string>& types) {
    fs::create_directories(outputDir);
    string path = outputDir + "\\" + baseName + ".java";
    ofstream writer(path);

    writer << "package com.craftinginterpreters.lox;\n\n";
    writer << "import java.util.List;\n\n";
    writer << "abstract class " << baseName << " {\n\n";

    defineVisitor(writer, baseName, types);

    // Define each AST type
    for (const string& type : types) {
        size_t colon = type.find(':');
        string className = type.substr(0, colon);
        className.erase(remove(className.begin(), className.end(), ' '), className.end());
        string fields = type.substr(colon + 1);
        fields.erase(0, fields.find_first_not_of(" "));
        defineType(writer, baseName, className, fields);
    }

    // Base accept method
    writer << "\n    abstract <R> R accept(Visitor<R> visitor);\n";
    writer << "}\n";
    writer.close();
}

void defineType(ofstream& writer, const string& baseName, const string& className, const string& fieldList) {
    writer << "\n    static class " << className << " extends " << baseName << " {\n";

    // Constructor
    writer << "        " << className << "(" << fieldList << ") {\n";
    stringstream ss(fieldList);
    string field;
    while (getline(ss, field, ',')) {
        field.erase(0, field.find_first_not_of(' '));
        size_t space = field.find(' ');
        string name = field.substr(space + 1);
        writer << "            this." << name << " = " << name << ";\n";
    }
    writer << "        }\n";

    // Visitor pattern accept
    writer << "\n        @Override\n";
    writer << "        <R> R accept(Visitor<R> visitor) {\n";
    writer << "            return visitor.visit" << className << baseName << "(this);\n";
    writer << "        }\n";

    // Fields
    writer << "\n";
    ss.clear();
    ss.str(fieldList);
    while (getline(ss, field, ',')) {
        field.erase(0, field.find_first_not_of(' '));
        writer << "        final " << field << ";\n";
    }

    writer << "    }\n";
}

void defineVisitor(ofstream& writer, const string& baseName, const vector<string>& types) {
    writer << "    interface Visitor<R> {\n";

    for (const string& type : types) {
        size_t colon = type.find(':');
        string typeName = type.substr(0, colon);
        typeName.erase(remove(typeName.begin(), typeName.end(), ' '), typeName.end());
        writer << "        R visit" << typeName << baseName << "(" << typeName << " " << baseName << ");\n";
    }

    writer << "    }\n";
}
